src/mapstorage/XmlMapStorage.cpp

```diff
@@ -0,0 +1,462 @@
+    saveMarkers(stream, markerList);
+    // write selected room x,y,z
+    saveCoordinate(stream, "position", m_mapData.getPosition());
+
+    stream.writeEndElement(); // end map
+
+    log("Writing data finished.");
+
+    m_mapData.unsetDataChanged();
+    emit sig_onDataSaved();
+
+    return true;
+}
+
+void XmlMapStorage::saveRooms(QXmlStreamWriter &stream,
+                              bool baseMapOnly,
+                              const ConstRoomList &roomList)
+{
+    ProgressCounter &progressCounter = getProgressCounter();
+    BaseMapSaveFilter filter;
+    if (baseMapOnly) {
+        filter.setMapData(&m_mapData);
+        progressCounter.increaseTotalStepsBy(filter.prepareCount());
+        filter.prepare(progressCounter);
+    }
+    auto saveOne = [&stream](const Room &room) { saveRoom(stream, room); };
+    for (const auto &pRoom : roomList) {
+        filter.visitRoom(deref(pRoom), baseMapOnly, saveOne);
+        progressCounter.step();
+    }
+}
+
+void XmlMapStorage::saveRoom(QXmlStreamWriter &stream, const Room &room)
+{
+    stream.writeStartElement("room");
+
+    stream.writeAttribute("id", QString("%1").arg(room.getId().asUint32()));
+    stream.writeAttribute("name", room.getName().toQString());
+    if (!room.isUpToDate()) {
+        stream.writeAttribute("uptodate", "false");
+    }
+    saveXmlElement(stream, "align", alignName(room.getAlignType()));
+    saveXmlElement(stream, "light", lightName(room.getLightType()));
+    saveXmlElement(stream, "portable", portableName(room.getPortableType()));
+    saveXmlElement(stream, "ridable", ridableName(room.getRidableType()));
+    saveXmlElement(stream, "sundeath", sundeathName(room.getSundeathType()));
+    saveXmlElement(stream, "terrain", terrainName(room.getTerrainType()));
+    saveCoordinate(stream, "coord", room.getPosition());
+    saveRoomLoadFlags(stream, room.getLoadFlags());
+    saveRoomMobFlags(stream, room.getMobFlags());
+
+    for (const ExitDirEnum dir : ALL_EXITS7) {
+        saveExit(stream, room.exit(dir), dir);
+    }
+    saveXmlElement(stream, "description", room.getDescription().toQString());
+    saveXmlElement(stream, "contents", room.getContents().toQString());
+    saveXmlElement(stream, "note", room.getNote().toQString());
+
+    stream.writeEndElement(); // end room
+}
+
+void XmlMapStorage::saveCoordinate(QXmlStreamWriter &stream,
+                                   const QString &name,
+                                   const Coordinate &pos)
+{
+    stream.writeStartElement(name);
+    stream.writeAttribute("x", QString("%1").arg(pos.x));
+    stream.writeAttribute("y", QString("%1").arg(pos.y));
+    stream.writeAttribute("z", QString("%1").arg(pos.z));
+    stream.writeEndElement(); // end coord
+}
+
+void XmlMapStorage::saveExit(QXmlStreamWriter &stream, const Exit &e, ExitDirEnum dir)
+{
+    if (!e.exitIsExit() || e.outIsEmpty()) {
+        return;
+    }
+    stream.writeStartElement("exit");
```

Comment:
Should there be a parent element called "exits" for all of the exits?

Reply:
Hrm, I do agree with the <rooms> idea. What if intead of <to> we have <exits> has a bunch of <exit> that can be redundant? There is some duplication but its still better than a ton of <to> elements.

For example:
```
<exits>
<exit dir="n" doorname="door" to="1">
<exit dir="n" doorname="door" to="2">
<exit dir="e" to="3">
</exits>
```

Reply:
Sure, I was just trying to suggest something that would optimize on disk space given you had expressed some concerns. I'm fine with the verbosity since it will also make diffs easier to interpret.

Reply:
As agreed (hoping I correctly understood your preference), I made "to" an element, i.e.
```xml
<exit dir="north">
  <to>3</to>
  <to>4</to>
</exit>
```
so now there's no need to store a comma-separated list of room ids in a single "to" attribute

Reply:
Looks good. Very few rooms have multiple exits so this might be a little verbose but it's not the end of the world.